src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs

```diff
@@ -4170,17 +4169,31 @@ private BoundExpression BindNullCoalescingAssignmentOperator(AssignmentExpressio
             ReportSuppressionIfNeeded(leftOperand, diagnostics);
             BoundExpression rightOperand = BindValue(node.Right, diagnostics, BindValueKind.RValue);
 
-            // If either operand is bad, bail out preventing more cascading errors
-            if (leftOperand.HasAnyErrors || rightOperand.HasAnyErrors)
+            TypeSymbol leftType = leftOperand.Type;
+
+            // If left operand is bad, bail out immediately
+            if (leftOperand.HasAnyErrors)
             {
                 leftOperand = BindToTypeForErrorRecovery(leftOperand);
                 rightOperand = BindToTypeForErrorRecovery(rightOperand);
                 return new BoundNullCoalescingAssignmentOperator(node, leftOperand, rightOperand, CreateErrorType(), hasErrors: true);
             }
+            // If right operand is bad, make sure conversion is in place for better error recovery
+            else if (rightOperand.HasAnyErrors)
+            {
+                var rightOperandTargetType = leftType switch
+                {
+                    { IsReferenceType: true } or { IsValueType: false, TypeKind: TypeKind.TypeParameter } => leftType,
+                    { IsValueType: true } when leftType.IsNullableType() => leftType.GetNullableUnderlyingType(),
+                    _ => CreateErrorType()
```

Comment:
Not sure we should be doing this conversion when there's an error type here.

Reply:
Some expressions (e.g. unconverted object expression) need conversion no matter what. I remember when I implemented this I tried to avoid this error type to error type conversion, but that triggered an assert later, because sometimes unconverted expressions where not handled correctly, e.g. `notNullableValueType ??= new()`. So I left it there

Reply:
Sure, but those should just go through our standard mechanisms for this, `BindToNaturalTypeForErrorRecovery` or whatever it's called.

Reply:
Tried locally. This broke 2 tests: `ErrorRecovery_CollectionExpression_NotNullableValueType` and `ErrorRecovery_TargetTypedInterpolatedString_NotNullableValueType`. In the first one _converted_ type changed from error type to `null`. In the second one _convrted_ type changed form error to `string`. While for the first case I am not sure which output is better, for the second one I prefer current PR output (error). What do you think?

Reply:
I probably prefer the change.

Reply:
> I probably prefer the change.

Sorry, which one? Original PR change (error) or your suggested change (string)? The latter should also be what a compiler from `main` currently outputs

Reply:
The latter.